Fix quota with federated sharing

PVince81 opened the pull request:
## Description
Fix quota with federated sharing and add tests for that.

- quota check needs to ignore -3 (unlimited) quota value (returned by federated shares with no quota)
- if unknown quota when checking a file on a federated share, check the parent
- when not checking the parent, uploading a file to a federated share through DAV upload should still properly forward the insufficient storage exception and others

## Related Issue
Fixes https://github.com/owncloud/core/issues/29169

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
Run federation_features/federated.feature:240 and quota.feature.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
Catch errors when setting quota in integration tests

Commit:
Fix quota handling with federated shares

Properly detect SPACE_UNLIMITED in quota check.
Improve error handling when receiving errors like 507 when uploading
files to a remote server.

PVince81 commented:
Note: I might want to move the DAV additions to a separate PR as I currently don't know how to properly automatically test this without commenting out the "check the parent instead" logic.

PVince81 commented:
- [ ] remove commented out code

At least this way we'll let the remote server decide in some cases whether there is enough quota when the quota is not retrievable. For that we need to properly forward the 507 error in the DAV code.

PVince81 commented:
- [ ] suspicious entries when running the tests:
```
[Thu Oct 12 16:45:52 2017] 127.0.0.1:51582 [201]: /public.php/webdav/testquota.txt-olddav-regular
[Thu Oct 12 16:45:53 2017] 127.0.0.1:51586 [207]: /public.php/webdav/testquota.txt-olddav-regular
[Thu Oct 12 16:45:53 2017] 127.0.0.1:51590 [207]: /public.php/webdav/
[Thu Oct 12 16:45:53 2017] Undefined index: size at /srv/www/htdocs/owncloud/lib/private/Files/View.php#1296
```

I didn't see this View warning on master, but also master doesn't have this one test.

PVince81 commented:
Why can't I catch this with ??
```diff
diff --git a/lib/private/Files/View.php b/lib/private/Files/View.php
index 3e4af7cf3d..9cb9bb482f 100644
--- a/lib/private/Files/View.php
+++ b/lib/private/Files/View.php
@@ -1294,6 +1294,10 @@ class View {
                try {
                        // if the file is not in the cache or needs to be updated, trigger the scanner and reload the data
                        if (!$data || $data['size'] === -1) {
+                               if (!!$data && !isset($data['size'])) {
+                                       $stack = json_encode(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 50));
+                                       \OCP\Util::writeLog('DEBUG', 'Size not set: Stack: ' . $stack, \OCP\Util::WARN);
+                               }
                                $this->lockFile($relativePath, ILockingProvider::LOCK_SHARED);

```

PVince81 commented:
Ok, caught its stack trace:
```
[Thu Oct 12 17:20:32 2017] Undefined index: size at /srv/www/htdocs/owncloud/lib/private/Files/View.php#1296
[Thu Oct 12 17:20:32 2017] $data is {"checksum":"","mimetype":"application\/octet-stream","parent":"2147483889"}
[Thu Oct 12 17:20:32 2017] Size not set: Stack: [{"file":"\/srv\/www\/htdocs\/owncloud\/lib\/private\/Files\/View.php","line":1351,"function":"getCacheEntry","class":"OC\\Files\\View","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/apps\/files_versions\/lib\/Storage.php","line":195,"function":"getFileInfo","class":"OC\\Files\\View","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/apps\/files_versions\/lib\/Hooks.php","line":60,"function":"store","class":"OCA\\Files_Versions\\Storage","type":"::"},{"file":"\/srv\/www\/htdocs\/owncloud\/lib\/private\/legacy\/hook.php","line":105,"function":"write_hook","class":"OCA\\Files_Versions\\Hooks","type":"::"},{"file":"\/srv\/www\/htdocs\/owncloud\/apps\/dav\/lib\/Connector\/Sabre\/File.php","line":283,"function":"emit","class":"OC_Hook","type":"::"},{"file":"\/srv\/www\/htdocs\/owncloud\/apps\/dav\/lib\/Connector\/Sabre\/File.php","line":189,"function":"emitPreHooks","class":"OCA\\DAV\\Connector\\Sabre\\File","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/apps\/dav\/lib\/Connector\/Sabre\/Directory.php","line":162,"function":"put","class":"OCA\\DAV\\Connector\\Sabre\\File","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php","line":1095,"function":"createFile","class":"OCA\\DAV\\Connector\\Sabre\\Directory","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/CorePlugin.php","line":525,"function":"createFile","class":"Sabre\\DAV\\Server","type":"->"},{"function":"httpPut","class":"Sabre\\DAV\\CorePlugin","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/event\/lib\/EventEmitterTrait.php","line":105,"function":"call_user_func_array"},{"file":"\/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php","line":479,"function":"emit","class":"Sabre\\Event\\EventEmitter","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php","line":254,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/apps\/dav\/lib\/Server.php","line":236,"function":"exec","class":"Sabre\\DAV\\Server","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/apps\/dav\/appinfo\/v2\/remote.php","line":31,"function":"exec","class":"OCA\\DAV\\Server","type":"->"},{"file":"\/srv\/www\/htdocs\/owncloud\/remote.php","line":175,"args":["\/srv\/www\/htdocs\/owncloud\/apps\/dav\/appinfo\/v2\/remote.php"],"function":"require_once"}]
```

to be formatted soon...

PVince81 commented:
If I remove the quota from the integration test, the warning disappears... Weird.

PVince81 commented:
also if I set the quota high enough in the test, the warning disappears as well.

PVince81 commented:
If I revert every code change and keep the test, the error appears... so this existed before.

PVince81 commented:
I've raised https://github.com/owncloud/core/issues/29245 to explain and analyze this bug. It seems unrelated to this PR. It's just that adding the new tests brought it into view.

PVince81 commented:
Ok, now that we know the warning is unrelated, let's focus on finishing this.

jvillafanez reviewed (commented):
Code looks good, but tests are failing

PVince81 commented:
Relevant failing tests:
```
19:19:53     /var/lib/jenkins/workspace/owncloud-core_core_PR-29220-7DOBZKU5OFHAVV4BVOWUOKNOOB2WEMRXAXPN3EVSUD4HDB6XAU2Q/tests/integration/features/external-storage.feature:63
19:19:53     /var/lib/jenkins/workspace/owncloud-core_core_PR-29220-7DOBZKU5OFHAVV4BVOWUOKNOOB2WEMRXAXPN3EVSUD4HDB6XAU2Q/tests/integration/features/sharing-v1.feature:597
```

PVince81 commented:
Ha... failing tests are failing because I added an extra check to assert that setting the quota worked. And apparently here it didn't but the test still passed.

Commit:
Fix integration tests to properly set quota in some cases

PVince81 commented:
stable10: https://github.com/owncloud/core/pull/29259

Commit:
Update QuotaPluginTest for unlimited quota values

Commit:
Added WebdavClientService in core + unit tests for Storage\DAV class

Added tests for all FS methods of the Storage\DAV class and also fixed
some related issues.

PVince81 commented:
I went through great lengths to make codecov happy: https://github.com/owncloud/core/pull/29220/commits/5b5689c0b64110afa43a2ce6407c5d0b6975cca2

- make the DAV class testable by introducing a WebdavClientService  and IWebdavClientService in core to be able to mock an instance of the Sabre Client
- add tests for the exceptions

Now waiting to see the new coverage and I'll continue adding tests for that.

PVince81 commented:
Let's watch the test coverage for the DAV class increase now!

PVince81 commented:
@jvillafanez @DeepDiver1975 please review again. To make codecov happy I had to go to great length and bootstrap unit tests for the Storage\DAV class. In my way of doing this I decided to cover all methods and discovered some small issues on the way which I fixed right away, some others which I will raise later.

PVince81 commented on tests/lib/Files/Storage/DavTest.php:
```diff
@@ -0,0 +1,1320 @@
+ * ownCloud
+ *
+ * @author Robin Appelman
+ * @copyright 2012 Robin Appelman icewind@owncloud.com
```
going to fix that right away...

PVince81 commented:
Rebased and squashed the DAV test (+fix) commits together into one.

DAV.php coverage is now 88%

jvillafanez commented on lib/private/Http/Client/WebdavClientService.php:
```diff
@@ -0,0 +1,80 @@
+	/**
+	 * Instantiate new Sabre client
+	 *
+	 * @param array $settings
```
Valid settings should be known both here and in the interface

DeepDiver1975 commented on lib/private/Http/Client/WebdavClientService.php:
```diff
@@ -0,0 +1,80 @@
+ *
+ * @package OC\Http
+ */
+class WebdavClientService implements IWebdavClientService {
```
Should be named WebDav

DeepDiver1975 commented on tests/lib/Files/Storage/DavTest.php:
```diff
@@ -0,0 +1,1319 @@
+		$this->timeFactory = $this->createMock(ITimeFactory::class);
+		$this->overwriteService('TimeFactory', $this->timeFactory);
+
+		$this->httpClient = $this->createMock(IClient::class);
```
missing use of IClient

PVince81 replied:
it's there: https://github.com/owncloud/core/pull/29220/files/4ba6cda1191234542d58f0b5be415392adca214b#diff-52c6f7ef7e46328fff29098af83d4b3cR26

PVince81 commented:
I've copied the Sabre client settings PHPDoc into the interface and WebdavClientService.

Since CI was all green and the PHPDoc will not affect it, I'm merging this directly.

DeepDiver1975 commented:
I was still right in the middle of review ....

PVince81 commented:
stable10: https://github.com/owncloud/core/pull/29325

PVince81 commented:
@DeepDiver1975 sorry, please continue then. I'll adjust in a subsequent PR.

DeepDiver1975 commented on tests/lib/Files/Storage/DavTest.php:
```diff
@@ -0,0 +1,1319 @@
+	}
+
+	/**
+	 * @expectedException OCP\Files\StorageNotAvailableException
```
namespace looks strange

![bildschirmfoto von 2017-10-23 12-20-52](https://user-images.githubusercontent.com/1005065/31884214-b904f122-b7ec-11e7-90bd-882bc64612ac.png)

DeepDiver1975 commented on tests/lib/Files/Storage/DavTest.php:
```diff
@@ -0,0 +1,1319 @@
+	private $webdavClientService;
+
+	/**
+	 * @var Client
```
please add | PHPUnit_Framework_MockObject_MockObject - this will fix issues with calling mock methods on these objects

![bildschirmfoto von 2017-10-23 12-23-03](https://user-images.githubusercontent.com/1005065/31884310-f0e71ffc-b7ec-11e7-8a90-ab0dbbe6fd85.png)

DeepDiver1975 commented:
> @DeepDiver1975 sorry, please continue then. I'll adjust in a subsequent PR.

done

PVince81 commented:
Follow up PR for the review items: https://github.com/owncloud/core/pull/29327

PVince81 commented:
regression: https://github.com/owncloud/core/pull/29422
